Porter.elm

```diff
@@ -73,24 +99,95 @@ encode encodeReq id msg =
 {-| Module messages.
 -}
 type Msg req res msg
-    = SendWithNextId (res -> msg) req
+    = SendWithNextId (FullRequest req res msg)
     | Receive Encode.Value
 
 
+{-| Internal type used by requests that have a response handler.
+-}
+type FullRequest req res msg
+    = FullRequest req (List (res -> Request req res)) (res -> msg)
+
+
+{-| Opaque type of a 'request'. Use the `request` function to create one,
+chain them using `andThen` and finally send it using `sendRequest`.
+-}
+type Request req res
+    = Request req (List (res -> Request req res))
+
+
 {-| Subscribe to messages from ports.
 -}
-subscriptions : Config req res msg -> Sub (Msg req res msg)
+subscriptions : Config req res msg -> Sub msg
 subscriptions config =
     config.incomingPort Receive
+        |> Sub.map config.porterMsg
 
 
 {-| Initiate a message send.
 -}
-send : (res -> msg) -> req -> Cmd (Msg req res msg)
-send responseHandler msg =
-    SendWithNextId responseHandler msg
+send : Config req res msg -> (res -> msg) -> req -> Cmd msg
+send config responseHandler request =
+    runSendRequest config (FullRequest request [] responseHandler)
+
+
+{-| Starts a request that can be sent at a later time using `sendRequest`,
+and that can be combined using `andThen`.
+-}
+request : req -> Request req res
+request req =
+    Request req []
+
+
+{-| Chains two Porter requests together:
+Run a second one right away when the first returns using its result in the request.
+-}
+andThen : (res -> Request req res) -> Request req res -> Request req res
+andThen reqfun (Request initial_req reqfuns) =
+    Request initial_req (reqfun :: reqfuns)
+
+
+{-| Transforms a request
+-}
+map : (res -> req) -> Request req res -> Request req res
+map func reqA =
+    reqA
+        |> andThen (\a -> request (func a))
+
+
+{-| Run a request using the results of two earlier requests.
+-}
+map2 : (res -> res -> req) -> Request req res -> Request req res -> Request req res
+map2 func reqA reqB =
+    reqA
+        |> andThen (\a -> reqB
+        |> andThen (\b -> request (func a b)))
+
+
+{-| -}
+map3 : (res -> res -> res -> req) -> Request req res -> Request req res -> Request req res -> Request req res
+map3 func reqA reqB reqC =
+    reqA
+      |> andThen (\a -> reqB
+      |> andThen (\b -> reqC
+      |> andThen (\c -> request (func a b c))))
+
+
+{-| Sends a request earlier started using `request`.
+-}
+sendRequest : Config req res msg -> (res -> msg) -> Request req res -> Cmd msg
```

Comment:
I would vouch for calling this one `send` instead of `sendRequest`, for consistency with `Http.send`. I wonder if we should keep the other `send` at all - feels like it would make the API confusing, and if people really want the shorthand, they can just define it as a `send << request` using the new API, assuming we go for the rename here. What do you think?

Reply:
Good question! Looking at `elm-lang/core`'s `Task` and `elm-lang/http`'s `Http` modules, I think that indeed having both makes the API confusing.

Both of those modules have a creation call (Http.get/Http.post/Http.request, Task.succeed/Task.fail) that is separate from the performing call (Http.send, Task.perform).

So: Yes, let's go for the rename here.